### DISCUSSION TITLE
Imports inside application factory

### DISCUSSION THREAD

**Discussion by fatadel** [pallets/flask#4275] [Q&A] [answered]
Hi everyone!

I am curious what's the rationale behind using imports inside the application factory function instead of regular top-level imports? See https://github.com/pallets/flask/blob/2f0c62f5e6e290843f03c1fa70817c7a3c7fd661/examples/tutorial/flaskr/__init__.py#L39

**Comment by fatadel** [comment]
I think I got it. Is it because when importing from a package it automatically invokes the init file? If so, I am curious why then importing at the top-level doesn't break anything?

**Comment by miguelgrinberg** [comment, marked as answer]
Imports inside the factory function are one of the techniques you can use to avoid circular import errors.

**Reply by fatadel** [reply]
But why putting the imports at top-level didn't cause any errors for me (i.e. with circular imports)?

**Reply by miguelgrinberg** [reply]
I don't know. You asked a general question and my answer is also given in general. Maybe this specific application you are looking at does not have any circular imports in its current state. Or maybe it did at some point during its development and that's the reason the imports were moved inside the app factory. If you prefer to have these imports at the top of the file then by all means you should have them there, there is no hard rule on this. But the fact remains, for a lot of applications this technique prevents the import system from looping on itself.

**Reply by davidism** [reply]
If you were to upgrade to Flask-SQLAlchemy, for example, it would be easier to introduce a circular import. A common pattern is to define a `db = SQLAlchemy()` object in the same file as `create_app`, and import that in other places. If you imported a blueprint at the top that also imported the `db` object, error. Moving the blueprint imports to right before they're registered keeps the rather generic names out of the global namespace and prevents any circular imports.

**Reply by fatadel** [reply]
Okay, so the concern is only about circular imports. Thank you very much, everyone! 😊

**Reply by fatadel** [reply]
@miguelgrinberg Sorry, if I misled you, but the question was not general. I've referred to a specific application from this repo (even a specific line in that application).
